File: python/context_engineering/session.py
"""
Differential Context Sessions

Stateful session manager that tracks what was sent in previous requests
and computes minimal diffs between turns. Combined with cache-topology
packing, this maximizes prefix cache reuse across API calls.

Research: ACE framework showed 83.6% lower token costs with incremental
delta updates compared to full recomputation baselines.
"""

from __future__ import annotations

import hashlib
import threading
from dataclasses import dataclass
from typing import Any, Dict, List, Optional

from .core import Budget, ContextItem, estimate_tokens, pack


@dataclass
class ManifestEntry:
    """A manifest entry tracking an item's position and content hash."""

    id: str
    content_hash: str
    tokens: int
    position: int


@dataclass
class SessionDelta:
    """Delta between two session states."""

    added: List[ContextItem]
    removed_ids: List[str]
    changed: List[ContextItem]
    kept_count: int
    delta_tokens: int
    reusable_tokens: int
    reuse_ratio: float


@dataclass
class SessionPack:
    """Result from session compile."""

    selected: List[ContextItem]
    dropped: List[ContextItem]
    total_tokens: int
    delta: Optional[SessionDelta]
    cache_key: str
    compile_count: int


def _quick_hash(s: str) -> str:
    """SHA-256 based hash for manifests and cache keys.

    Returns the first 16 hex characters for a compact but
    collision-resistant key.
    """
    return hashlib.sha256(s.encode()).hexdigest()[:16]


def _unchanged_prefix(
    previous: List[ManifestEntry], current: List[ManifestEntry]
) -> List[ManifestEntry]:
    prefix: List[ManifestEntry] = []
    for prev_entry, curr_entry in zip(previous, current):
        if prev_entry.id != curr_entry.id:
            break
        if prev_entry.content_hash != curr_entry.content_hash:
            break
        prefix.append(curr_entry)
    return prefix


class ContextSession:
    """A stateful context session that tracks changes between compiles.

    On each compile(), it packs the current items and computes the diff
    from the previous compile. This enables:
    - Tracking how much context actually changes between turns
    - Computing cache reuse ratios for cost estimation
    - Identifying stable prefixes that benefit from KV cache

    Example:
        session = create_session(Budget(max_tokens=8000))

        # Turn 1
        session.set_items([system_prompt, doc1, doc2, query1])
        r1 = session.compile()
        # r1.delta is None (first compile)

        # Turn 2 — only query changed
        session.set_items([system_prompt, doc1, doc2, query2])
        r2 = session.compile()
        # r2.delta.reuse_ratio ~ 0.85 (most tokens cached)
    """

    def __init__(self, budget: Budget, pack_options: Optional[Dict[str, Any]] = None):
        self._budget = budget
        self._pack_options = pack_options or {}
        self._current_items: List[ContextItem] = []
        self._previous_manifest: List[ManifestEntry] = []
        self._compile_count = 0
        self._lock = threading.Lock()

    def set_items(self, items: List[ContextItem]) -> None:
        """Replace the current item set."""
        with self._lock:
            self._current_items = list(items)

    def add_items(self, items: List[ContextItem]) -> None:
        """Add items to the current set (deduplicates by id)."""
        with self._lock:
            existing = {i.id: i for i in self._current_items}
            for item in items:
                existing[item.id] = item
            self._current_items = list(existing.values())

    def remove_items(self, ids: List[str]) -> None:
        """Remove items by ID."""
        with self._lock:
            remove_set = set(ids)
            self._current_items = [i for i in self._current_items if i.id not in remove_set]

    def compile(self) -> SessionPack:
        """Compile the current context, computing delta from previous."""
        with self._lock:
            # Pack current items
            packed = pack(self._current_items, self._budget, **self._pack_options)

            # Build manifest for current compile
            current_manifest: List[ManifestEntry] = []
            for i, item in enumerate(packed.selected):
                tokens = item.tokens or estimate_tokens(item.content)
                current_manifest.append(
                    ManifestEntry(
                        id=item.id,
                        content_hash=_quick_hash(item.content),
                        tokens=tokens,
                        position=i,
                    )
                )

            # Compute delta from previous
            delta: Optional[SessionDelta] = None

            # Compute unchanged prefix once (used for both delta and cache key).
            unchanged_prefix = _unchanged_prefix(self._previous_manifest, current_manifest)

            if self._compile_count > 0:
                prev_map = {e.id: e for e in self._previous_manifest}
                curr_map = {e.id: e for e in current_manifest}
                reusable_prefix_ids = {entry.id for entry in unchanged_prefix}

                added: List[ContextItem] = []
                changed: List[ContextItem] = []
                removed_ids: List[str] = []
                kept_count = 0
                added_tokens = 0
                removed_tokens = 0
                changed_tokens = 0
                reusable_tokens = 0

                # Build a map from id -> item for O(1) lookups
                selected_map = {i.id: i for i in packed.selected}

                # Find added and changed
                for entry in current_manifest:
                    prev = prev_map.get(entry.id)
                    if prev is None:
                        item = selected_map.get(entry.id)
                        if item:
                            added.append(item)
                        added_tokens += entry.tokens
                    elif prev.content_hash != entry.content_hash:
                        item = selected_map.get(entry.id)
                        if item:
                            changed.append(item)
                        changed_tokens += entry.tokens
                    else:
                        kept_count += 1
                        if entry.id in reusable_prefix_ids:
                            reusable_tokens += entry.tokens

                # Find removed
                for entry in self._previous_manifest:
                    if entry.id not in curr_map:
                        removed_ids.append(entry.id)
                        removed_tokens += entry.tokens

                delta_tokens = added_tokens + removed_tokens + changed_tokens
                total_prev = sum(e.tokens for e in self._previous_manifest)

                delta = SessionDelta(
                    added=added,
                    removed_ids=removed_ids,
                    changed=changed,
                    kept_count=kept_count,
                    delta_tokens=delta_tokens,
                    reusable_tokens=reusable_tokens,
                    reuse_ratio=round(reusable_tokens / total_prev, 3) if total_prev > 0 else 0,
                )

            # Generate cache key from unchanged prefix
            cache_key = _quick_hash(
                ",".join(
                    f"{entry.position}:{entry.id}:{entry.content_hash}"
                    for entry in unchanged_prefix
                )
                or "empty"
            )

            # Update state for next compile
            self._previous_manifest = current_manifest
            self._compile_count += 1

            return SessionPack(
                selected=list(packed.selected),
                dropped=list(packed.dropped),
                total_tokens=packed.total_tokens,
                delta=delta,
                cache_key=cache_key,
                compile_count=self._compile_count,
            )

    def item_count(self) -> int:
        """Get the current item count (before packing)."""
        with self._lock:
            return len(self._current_items)

    def get_compile_count(self) -> int:
        """Get the number of compiles performed."""
        with self._lock:
            return self._compile_count

    def clear(self) -> None:
        """Reset the session."""
        with self._lock:
            self._current_items = []
            self._previous_manifest = []
            self._compile_count = 0


def create_session(
    budget: Budget,
    pack_options: Optional[Dict[str, Any]] = None,
) -> ContextSession:
    """Create a new context session.

    Args:
        budget: Token budget for packing
        pack_options: Default pack options

    Returns:
        A ContextSession instance
    """
    return ContextSession(budget, pack_options)

File: packages/ce-core/src/session.ts
/**
 * Differential Context Sessions
 *
 * Stateful session manager that tracks what was sent in previous requests
 * and computes minimal diffs between turns. Combined with cache-topology
 * packing, this maximizes prefix cache reuse across API calls.
 *
 * Research: ACE framework showed 83.6% lower token costs with incremental
 * delta updates compared to full recomputation baselines.
 */

import type { Budget, ContextItem, PackOptions } from "./types.js";
import { pack } from "./pack.js";
import { estimateTokens } from "./estimate.js";
import { hash64 } from "./hash.js";

/**
 * A manifest entry tracking an item's position and content hash.
 */
interface ManifestEntry {
  id: string;
  contentHash: string;
  tokens: number;
  position: number;
}

/**
 * Delta between two session states.
 */
export interface SessionDelta {
  /** Items added since last compile */
  added: ContextItem[];
  /** Item IDs removed since last compile */
  removedIds: string[];
  /** Items whose content changed */
  changed: ContextItem[];
  /** Number of items unchanged */
  keptCount: number;
  /** Tokens that changed (added + removed + changed) */
  deltaTokens: number;
  /** Tokens reusable from the previous request's KV cache */
  reusableTokens: number;
  /** Fraction of tokens reusable (0-1) */
  reuseRatio: number;
}

/**
 * Result from session compile.
 */
export interface SessionPack {
  /** All selected items for the current request */
  selected: ContextItem[];
  /** Items that were dropped */
  dropped: ContextItem[];
  /** Total tokens in the compiled context */
  totalTokens: number;
  /** Delta from the previous compile (null on first compile) */
  delta: SessionDelta | null;
  /** Stable cache key (hash of unchanged prefix items) */
  cacheKey: string;
  /** Cumulative compile count */
  compileCount: number;
}

function unchangedPrefix(
  previous: ManifestEntry[],
  current: ManifestEntry[]
): ManifestEntry[] {
  const prefix: ManifestEntry[] = [];
  const maxLength = Math.min(previous.length, current.length);
  for (let i = 0; i < maxLength; i++) {
    const prev = previous[i];
    const curr = current[i];
    if (prev.id !== curr.id || prev.contentHash !== curr.contentHash) break;
    prefix.push(curr);
  }
  return prefix;
}

/**
 * A stateful context session that tracks changes between compiles.
 *
 * On each compile(), it packs the current items and computes the diff
 * from the previous compile. This enables:
 * - Tracking how much context actually changes between turns
 * - Computing cache reuse ratios for cost estimation
 * - Identifying stable prefixes that benefit from KV cache
 *
 * @example
 * ```ts
 * const session = createSession({ maxTokens: 8000 });
 *
 * // Turn 1
 * session.setItems([systemPrompt, doc1, doc2, query1]);
 * const r1 = session.compile();
 * // r1.delta is null (first compile)
 *
 * // Turn 2 — only query changed
 * session.setItems([systemPrompt, doc1, doc2, query2]);
 * const r2 = session.compile();
 * // r2.delta.reuseRatio ≈ 0.85 (most tokens cached)
 * ```
 */
export interface ContextSession {
  /** Replace the current item set */
  setItems(items: ContextItem[]): void;
  /** Add items to the current set */
  addItems(items: ContextItem[]): void;
  /** Remove items by ID */
  removeItems(ids: string[]): void;
  /** Compile the current context, computing delta from previous */
  compile(options?: PackOptions): SessionPack;
  /** Get the current item count (before packing) */
  itemCount(): number;
  /** Get the number of compiles performed */
  getCompileCount(): number;
  /** Reset the session */
  clear(): void;
}

export interface SessionOptions {
  /** Token budget */
  budget: Budget;
  /** Default pack options */
  packOptions?: PackOptions;
}

/**
 * Create a stateful context session that tracks changes between compiles.
 *
 * @param options - Session configuration including token budget and default pack options
 * @returns A ContextSession instance
 */
export function createSession(options: SessionOptions): ContextSession {
  const budget = options.budget;
  const defaultPackOptions = options.packOptions ?? {};

  let currentItems: ContextItem[] = [];
  let previousManifest: ManifestEntry[] = [];
  let compileCount = 0;

  function setItems(items: ContextItem[]): void {
    currentItems = [...items];
  }

  function addItems(items: ContextItem[]): void {
    // Deduplicate by id — new items override existing
    const existing = new Map(currentItems.map(i => [i.id, i]));
    for (const item of items) {
      existing.set(item.id, item);
    }
    currentItems = Array.from(existing.values());
  }

  function removeItems(ids: string[]): void {
    const removeSet = new Set(ids);
    currentItems = currentItems.filter(i => !removeSet.has(i.id));
  }

  function compile(packOptions?: PackOptions): SessionPack {
    const opts = { ...defaultPackOptions, ...packOptions };
    const estimator = opts.tokenEstimator;

    // Pack current items
    const packed = pack(currentItems, budget, opts);

    // Build manifest for current compile
    const currentManifest: ManifestEntry[] = packed.selected.map((item, i) => ({
      id: item.id,
      contentHash: hash64(item.content),
      tokens: item.tokens ?? estimateTokens(item.content, { estimator }),
      position: i,
    }));

    // Compute delta from previous
    let delta: SessionDelta | null = null;

    if (compileCount > 0) {
      const prevMap = new Map(previousManifest.map(e => [e.id, e]));
      const currMap = new Map(currentManifest.map(e => [e.id, e]));
      const reusablePrefix = unchangedPrefix(previousManifest, currentManifest);
      const reusablePrefixIds = new Set(reusablePrefix.map(entry => entry.id));

      const added: ContextItem[] = [];
      const changed: ContextItem[] = [];
      const removedIds: string[] = [];
      let keptCount = 0;
      let addedTokens = 0;
      let removedTokens = 0;
      let changedTokens = 0;
      let reusableTokens = 0;

      // Build a map from id -> item for O(1) lookups
      const selectedMap = new Map(packed.selected.map(i => [i.id, i]));

      // Find added and changed
      for (const entry of currentManifest) {
        const prev = prevMap.get(entry.id);
        if (!prev) {
          // New item
          const item = selectedMap.get(entry.id);
          if (item) added.push(item);
          addedTokens += entry.tokens;
        } else if (prev.contentHash !== entry.contentHash) {
          // Content changed
          const item = selectedMap.get(entry.id);
          if (item) changed.push(item);
          changedTokens += entry.tokens;
        } else {
          // Unchanged
          keptCount++;
          if (reusablePrefixIds.has(entry.id)) {
            reusableTokens += entry.tokens;
          }
        }
      }

      // Find removed
      for (const entry of previousManifest) {
        if (!currMap.has(entry.id)) {
          removedIds.push(entry.id);
          removedTokens += entry.tokens;
        }
      }

      const deltaTokens = addedTokens + removedTokens + changedTokens;
      const totalPrev = previousManifest.reduce((s, e) => s + e.tokens, 0);

      delta = {
        added,
        removedIds,
        changed,
        keptCount,
        deltaTokens,
        reusableTokens,
        reuseRatio:
          totalPrev > 0
            ? Math.round((reusableTokens / totalPrev) * 1000) / 1000
            : 0,
      };
    }

    // Generate cache key from unchanged items
    const prefix = unchangedPrefix(previousManifest, currentManifest);
    const cacheKey = hash64(
      prefix
        .map(entry => `${entry.position}:${entry.id}:${entry.contentHash}`)
        .join(",") || "empty"
    );

    // Update state for next compile
    previousManifest = currentManifest;
    compileCount++;

    return {
      selected: packed.selected,
      dropped: packed.dropped,
      totalTokens: packed.totalTokens,
      delta,
      cacheKey,
      compileCount,
    };
  }

  function itemCount(): number {
    return currentItems.length;
  }

  function getCompileCount(): number {
    return compileCount;
  }

  function clear(): void {
    currentItems = [];
    previousManifest = [];
    compileCount = 0;
  }

  return {
    setItems,
    addItems,
    removeItems,
    compile,
    itemCount,
    getCompileCount,
    clear,
  };
}
